llvm/include/llvm/CodeGen/AsmPrinter.h

```diff
@@ -905,6 +921,78 @@ class AsmPrinter : public MachineFunctionPass {
   }
 };
 
+class AsmPrinterInitializePass
+    : public PassInfoMixin<AsmPrinterInitializePass> {
+  std::shared_ptr<AsmPrinter> Printer;
```

Comment:
There should be no reason to use shared_ptr instead of unique_ptr

Reply:
Currently these three passes are tightly coupled, `AsmPrinterInitializePass ` may modify some states in `AsmPrinter`. Also, they must share the same streamer, which is used by all three passes.

Reply:
I would expect the state to move from one pass to the next. It does not need atomic reference counting

Reply:
What if the first pass owned the pointer, the next 2 depend on it, and the second 2 only use a non-smart reference?